python/vespa/vespa/package.py

```diff
@@ -468,3 +490,289 @@ def deploy(self, disk_folder: str, container_memory: str = "4G"):
+            **headers,
+        }
+
+        body.seek(0)
+        self.connection.request(method, path, body, headers)
+        with self.connection.getresponse() as response:
+            parsed = json.load(response)
+            if response.status != 200:
+                raise RuntimeError(
+                    "Status code "
+                    + str(response.status)
+                    + " doing "
+                    + method
+                    + " at "
+                    + url
+                    + ":\n"
+                    + parsed["message"]
+                )
+            return parsed
+
+    def get_dev_region(self) -> str:
+        return self.request("GET", "/zone/v1/environment/dev/default")["name"]
+
+    def get_endpoint(
+        self, instance: str, region: str, application_package_name: str
+    ) -> str:
+        endpoints = self.request(
+            "GET",
+            "/application/v4/tenant/{}/application/{}/instance/{}/environment/dev/region/{}".format(
+                self.tenant, self.application, instance, region
+            ),
+        )["endpoints"]
+        container_url = [
+            endpoint["url"]
+            for endpoint in endpoints
+            if endpoint["cluster"] == "{}_container".format(application_package_name)
+        ]
+        if not container_url:
+            raise RuntimeError("No endpoints found for container 'test_app_container'")
+        return container_url[0]
+
+    def to_application_zip(self) -> BytesIO:
+        buffer = BytesIO()
+        with zipfile.ZipFile(buffer, "a") as zip_archive:
+            zip_archive.writestr(
+                "application/schemas/{}.sd".format(
+                    self.application_package.schema.name
+                ),
+                self.application_package.schema_to_text,
+            )
+            zip_archive.writestr(
+                "application/services.xml", self.application_package.services_to_text
+            )
+            zip_archive.writestr(
+                "application/security/clients.pem",
+                self.data_certificate.public_bytes(serialization.Encoding.PEM),
+            )
+
+        return buffer
+
+    def start_deployment(self, instance: str, job: str, disk_folder: str) -> int:
+        deploy_path = (
+            "/application/v4/tenant/{}/application/{}/instance/{}/deploy/{}".format(
+                self.tenant, self.application, instance, job
+            )
+        )
+
+        application_zip_bytes = self.to_application_zip()
+
+        self.write_private_key_and_cert(
+            self.data_key, self.data_certificate, disk_folder
+        )
+        with open(os.path.join(disk_folder, "application.zip"), "wb") as zipfile:
+            zipfile.write(application_zip_bytes.getvalue())
+
+        response = self.request(
+            "POST",
+            deploy_path,
+            application_zip_bytes,
+            {"Content-Type": "application/zip"},
+        )
+        print(response["message"])
```

Comment:
No, it contains a message intended for the user.

Reply:
Wouldn't it be better to return that value from the method instead? Library code printing stuff is a bit iffy. I know there's a contention here between regular library code and interactive notebooks, but it would be nice to see if there was a better way.

Reply:
Yes, moving it up one level is a good idea.

Reply:
I don't think this was a left over. The idea is to print the deployment messages so that the user knows which stage he is. This `response["message"]` contains the following
```
Deployment started in run 6 of dev-aws-us-east-1c for vespa-team.ms-marco.from-notebook. This may take about 15 minutes the first time.
```
The entire message including the updates looks something like:
```
Deployment started in run 6 of dev-aws-us-east-1c for vespa-team.ms-marco.from-notebook. This may take about 15 minutes the first time.
INFO    [08:08:03]  Deploying platform version 7.276.8 and application version unknown ...
INFO    [08:08:04]  No services requiring restart.
INFO    [08:08:04]  Deployment successful.
INFO    [08:08:05]  Session 13618 for tenant 'vespa-team' prepared and activated.
INFO    [08:08:05]  ######## Details for all nodes ########
INFO    [08:08:05]  h711a.dev.aws-us-east-1c.vespa-external.aws.oath.cloud: expected to be UP
INFO    [08:08:05]  --- platform docker.ouroath.com:4443/vespa/centos-tenant:7.276.8
INFO    [08:08:05]  --- container on port 4080 has config generation 13618, wanted is 13618
INFO    [08:08:05]  h712a.dev.aws-us-east-1c.vespa-external.aws.oath.cloud: expected to be UP
INFO    [08:08:05]  --- platform docker.ouroath.com:4443/vespa/centos-tenant:7.276.8
INFO    [08:08:05]  --- logserver-container on port 4080 has config generation 13618, wanted is 13618
INFO    [08:08:05]  h713a.dev.aws-us-east-1c.vespa-external.aws.oath.cloud: expected to be UP
INFO    [08:08:05]  --- platform docker.ouroath.com:4443/vespa/centos-tenant:7.276.8
INFO    [08:08:05]  --- container-clustercontroller on port 19050 has config generation 13618, wanted is 13618
INFO    [08:08:05]  --- storagenode on port 19102 has config generation 13618, wanted is 13618
INFO    [08:08:05]  --- searchnode on port 19107 has config generation 13618, wanted is 13618
INFO    [08:08:05]  --- distributor on port 19111 has config generation 13618, wanted is 13618
INFO    [08:08:05]  Found endpoints:
INFO    [08:08:05]  - dev.aws-us-east-1c
INFO    [08:08:05]   |-- https://msmarco-container.from-notebook.ms-marco.vespa-team.aws-us-east-1c.dev.public.vespa.oath.cloud/ (cluster 'msmarco_container')
INFO    [08:08:06]  Installation succeeded!
```

Reply:
@oyving Would you change it?

Reply:
I think it should be changed, but it might be over-engineering it. It's your call. But as a minimum I would have a `self.output` file-handle that you write these messages to (that default to `sys.stdout`). That way the output can at least be controlled by the user if they wish.

Reply:
Done.